### PR TITLE
Update to .net core 3.1.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

- Update sample apps to `netcoreapp3.1`
- Remove netfx targets from samples as they seem to interfere somehow with net core targets
- Pin SDK in global.json.

### CODE REVIEW

**Comment by MarchingCube** [comment]
Maybe we can bump benchmarks as well? Since I was always doing that manually anyway.

**Comment by MarchingCube** [comment]
It also would be great to multi target .net standard 2.1 as well in the main projects.

**Comment by grokys** [comment]
Ah yeah, I also need to bump unit tests.

Not sure about dual-targeting .net standard 2.1, what does everyone think? Is it likely to cause us any problems?

File: src/tools/Avalonia.Designer.HostApp/Avalonia.Designer.HostApp.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>
+    <TargetFrameworks>net461;netcoreapp3.1</TargetFrameworks>
```
Comment (grokys) [inline src/tools/Avalonia.Designer.HostApp/Avalonia.Designer.HostApp.csproj]: Not 100% sure we want to change this?

**Comment by grokys** [comment]
Also `/Packages/Avalonia.csproj` multi-targets `netstandard2.0;net461;netcoreapp2.0` - not sure why it does this, do we want to change that to 3.1 too?

**Comment by grokys** [comment]
Ok, seems nuke is failing on .NET core 3.1, looks like we're going to have to upgrade it, but there have been a lot of breaking changes between 0.12 and 0.24, so this looks like it's not going to be as straightforward as I'd hoped. Marking this PR WIP for now.

File: src/tools/Avalonia.Designer.HostApp/Avalonia.Designer.HostApp.csproj (hunk 1 above)
Comment (kekekeks) [reply src/tools/Avalonia.Designer.HostApp/Avalonia.Designer.HostApp.csproj]: Host app TFM should match the TFM of Avalonia.DesktopRuntime. Which is not changed by this PR.

**Comment by kekekeks** [comment]
Could we use both .NET Core 3.1 and the oldest supported one instead? We are targeting `netcoreapp2.0` TFM anyway.

**Comment by grokys** [comment]
Ah yeah, i meant to change `Avalonia.DesktopRuntime` to 3.1 as well. Is there any reason to target 2.0 still?

**Comment by kekekeks** [comment]
OSX 10.12 is not supported by .NET Core 3.1. Not sure if we care about that.

**Comment by MarchingCube** [comment]
@kekekeks It looks like only 7% people use macOS 10.12 at the moment and it is shrinking (according to https://gs.statcounter.com/macos-version-market-share/desktop/worldwide).

@grokys It would be good to figure out .Net Standard 2.1 support soon, since it is blocking certain work (like improving our string tokenizers and providing more span based optimizations).

**Comment by Gillibald** [comment]
What exactly will perform better with the tokenizer under. Net Standard 2.1?

**Comment by MarchingCube** [comment]
@Gillibald Almost everything :) Since it currently works on substrings https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Utilities/StringTokenizer.cs#L43. And .Net Standard 2.1 adds `ReadOnlySpan<T>` based overloads for parsing, so we could do tokenization without allocations.

https://apisof.net/catalog/System.Double.TryParse(ReadOnlySpan%3CChar%3E,NumberStyles,IFormatProvider,Double)

**Comment by jmacato** [comment]
It'd be great if we could move all our targets to netstandard2.1 and netcoreapp3.1. so many perf stuff in it as @MarchingCube said.

**Comment by kekekeks** [comment]
netstandard2.1 is not supported by net4xx TFMs. Which makes it impossible to run our unit tests on Mono. That would be a road blocker for mobile platform support :-1:

**Comment by jkoritzinsky** [comment]
IIRC Mono does support netstandard2.1 but there’s just no good way with the standard tooling to do it.

**Comment by kekekeks** [comment]
@jkoritzinsky That's simply because there is no desktop Mono target framework moniker. It has to reuse net4xx while being a superset of that API.

**Comment by kekekeks** [comment]
Also, keep in mind, that while .NET 5 will have Mono as a runtime option, Xamarin platforms aren't migrating from the full mono codebase anytime soon.

**Comment by jkoritzinsky** [comment]
I don’t know if you saw the new TFM design spec for .NET 5, but Xamarin is planning on moving to a .NET 5 based TFM if all goes according to plan.

**Comment by kekekeks** [comment]
If I understand correctly, the BCL will still be based on `mono/mcs/class` rather than one from dotnet/runtime repo.

**Comment by jkoritzinsky** [comment]
Now that I’m not sure about. But considering a very significant portion of the BCL is shared, I wouldn’t worry about BCL differences being our problem. Our Mono-specific issues have almost always been runtime issues around either bugs or underdefined behavior.

**Comment by kekekeks** [comment]
Also, keep in mind that since we'd be shipping our NativeControlHost in the next release, embedding WPF/WinForms controls into Avalonia is now possible. With existing Avalonia into WPF/WinForms embedding, the drop of `net4xx` TFM support effectively removes a rather painless way of migrating existing applications, people will have to update their existing WPF code first instead of migrating to Avalonia directly.

**Comment by jkoritzinsky** [comment]
I think we should do the following:

Make Avalonia.dll a ref assembly for all of Avalonia that targets netstandard2.0 with its implementation assembly just type-forwarding. Make the rest of the assemblies implementation-only and multitargetting. That way we can have a unified API surface on all platforms, have “public” types we don’t expose to users, and use platform-specific APIs under the hood for performance.

**Comment by kekekeks** [comment]
>Make the rest of the assemblies implementation-only and multitargetting. 

That would conflict with how AppBuilder currently works. Right now it's only provided by `netcoreappXX` and `net4XX` TFMs. On mobile it's provided by backend-specific Avalonia.iOS and Avalonia.Android packages. It's not provided by `netstandard2.0` dll.

**Comment by jkoritzinsky** [comment]
AppBuilder is in the Avalonia.Desktop package right? I was referring to the stuff in the Avalonia package only. Sorry about not clarifying that.

**Comment by kekekeks** [comment]
AppBuilder in Avalonia package itself.

**Comment by kekekeks** [comment]
For now I'd suggest to write the list of things that we need from netstandard2.1 and see if those could be polyfilled. It's not like we need `Span` overloads for System.IO.Stream.

**Comment by ahopper** [comment]
I did try using a the span parse overloads on Geometry.Parse, this did reduce the number of allocations but actually had only a small speed increase, the underlying double parsing code is slow and dominates, probably because it has to allow for every possible formatting option.  There are possibly bigger gains by using custom number span parsing (compatible with net standard 2.0) or better still do the parsing in xamlil at build time.

**Comment by grokys** [comment]
Could we just `#if` stuff depending on the target framework? Or would that be too much of a hassle?

**Comment by jkoritzinsky** [comment]
I’m fine with using #if per TFM. I want the ref assembly though to ensure that our public surface area is consistent.

**Comment by kekekeks** [comment]
>the underlying double parsing code is slow and dominates, probably because it has to allow for every possible formatting option

I guess we should just use some specialized parser.

**Comment by matkoch** [comment]
@grokys @kekekeks please let me know if you need any help with updating the build.

**Comment by grokys** [comment]
Thanks @matkoch! Am I right in thinking that nuke 0.12 doesn't support .net core 3.1?

**Comment by matkoch** [comment]
I honestly don’t know about the compatibility implications between
netcoreapp2.0 and netcoreapp3.0 but I can at least say that we made the
switch with version 0.23 of nuke. do you really want to stay at 0.12?

**Comment by kekekeks** [comment]
I'd suggest to leave the Nuke-based project as is and update it in a separate PR.

**Comment by grokys** [comment]
I've merged in #3596 (which updates the nuke build to 0.24) and CI now passes. Unmarking as WIP.

**Review by MarchingCube — APPROVED** [review]
LGTM. Unless anybody has objections - we should merge this in and then deal with consequences if any arise :)

**Pull request merged**